README
This is the website i created for my wedding. It uses TiddlyWiki as the main framework, then i changed the layout, added plugins etc to make it a complete website. 
Some of the special features:
* Tiddlywiki based, with minimal php requirements (php was used to authenticate user name so i can edit and save the tiddlywiki online)
* Major customization of Tiddlywiki
* Integrated Tiddlywiki with a message board (the same one used for my website that used hand coded AJAX)
* Added flash player and built in music control

Links:
*Tiddlywiki - http://www.tiddlywiki.com/

board/CHANGELOG
changelog

new in 1.0.3:
- minor bugfixes
- install information

new in 1.0.2:
- emailaddress validation
- removed a bug which malformed the xml file when a message contained an '&'
- filelocking

new in 1.0.1
- some documentation
- examplefiles now include descriptive content


board.js

var req;
function importXMLDoc() {
    // branch for native XMLHttpRequest object
    if (window.XMLHttpRequest) {
        req = new XMLHttpRequest();
        req.onreadystatechange = processReqChange;
        req.open("GET", "board/data/board.xml", true);
        req.send(null);
    // branch for IE/Windows ActiveX version
    } else if (window.ActiveXObject) {
        isIE = true;
        req = new ActiveXObject("Microsoft.XMLHTTP");
        if (req) {
            req.onreadystatechange = processReqChange;
            req.open("GET", "board/data/board.xml", true);
            req.send();
        }
    }
}

// handle onreadystatechange event of req object
function processReqChange() {
    // only if req shows "loaded"
    if (req.readyState == 4) {
        // only if "OK"
        if (req.status == 200) {
            createMain();
         } else {
            alert("There was a problem retrieving the XML data:\n" +
                req.statusText);
         }
    }
}
 
function createMain()
{
	var xb = req.responseXML.getElementsByTagName('MESSAGE');
	var para = document.createElement('div');
        para.setAttribute('id', 'boardContent');
	var len = xb.length<20 ? xb.length:20;
        for (i=0; i<len; i++)
	{
	
		var j = 0;
		
		while (xb[i].childNodes[j].nodeType !=1) j++;
		var theDate = document.createTextNode(xb[i].childNodes[j].firstChild.nodeValue);
		j++;
		while (xb[i].childNodes[j].nodeType !=1) j++;
		var author = document.createTextNode(xb[i].childNodes[j].firstChild.nodeValue);
		var authemail = "mailto:";
		authemail += xb[i].childNodes[j].attributes.getNamedItem('email').nodeValue;
		j++;
		while (xb[i].childNodes[j].nodeType !=1) j++;
		var msgtext = document.createTextNode(xb[i].childNodes[j].firstChild.nodeValue);
		
		
		var newEl = document.createElement('TABLE');
		newEl.setAttribute('cellPadding',1);
		newEl.setAttribute('width', '95%');
		var tmp = document.createElement('TBODY');
		newEl.appendChild(tmp);
		para.appendChild(newEl);
		var row = document.createElement('TR');
		var container = document.createElement('TH');
		container.appendChild(author);
		row.appendChild(container);
		var container2 = document.createElement('TD');
		container2.appendChild(theDate);
                container2.setAttribute('id', 'msgBoardDate');
		container2.setAttribute('align', 'right');
		row.appendChild(container2);
		tmp.appendChild(row);

		var row2 = document.createElement('TR');
		var container3 = document.createElement('TD');
		container3.setAttribute('ColSpan', '2');
		container3.appendChild(msgtext);
		row2.appendChild(container3);
		tmp.appendChild(row2);
		para.appendChild(newEl);

        } 

        if ( document.getElementById('maincontent').hasChildNodes() ) {
                document.getElementById('maincontent').replaceChild(para, document.getElementById('maincontent').childNodes[0]);	
                
        }
        else
	        document.getElementById('maincontent').appendChild(para);	
}
importXMLDoc();


board/board.php
<?php
/****************************************
* XMB v1.0 (XmlMessageBoard)
*****************************************
* written by erwin vrolijk,
* erwin@redant.nl
*****************************************
* this script is released under the GPL
* http://www.gnu.org/copyleft/gpl.html
****************************************/

require("./scripts/conf.php");
require("./scripts/print.php");
require("./scripts/xmlparser.php");

header("Expires: Mon, 26 Jul 1997 05:00:00 GMT" );
header("Content-Type: text/xml; charset=utf-8");
header("Last-Modified: " . gmdate("D, d M Y H:i:s") . " GMT");
header("Cache-Control: no-store, no-cache, must-revalidate");
header("Cache-Control: post-check=0, pre-check=0", false);
header("Pragma: no-cache");
if (!isset($start)) {
	$start = 1;
}

print_from_file("./data/board.xml");

?>


board/scripts/add.php
<?php
/****************************************
* XMB v1.0 (XmlMessageBoard)
* http://xmboard.sourceforge.net
*****************************************
* written by erwin vrolijk,
* aierwin@users.sourceforge.net
*****************************************
* this script is released under the GPL
* http://www.gnu.org/copyleft/gpl.html
****************************************/

/* add.php

   requires conf.php
   requires print.php
*/

require("./conf.php");
require("./print.php");
/*
require_once('recaptchalib.php');
$privatekey = "6LdlXAAAAAAAAFpb741ZeEgK-ni7xvrNfXH5yNNa";
$resp = recaptcha_check_answer ($privatekey,
                                $_SERVER["REMOTE_ADDR"],
                                $_POST["recaptcha_challenge_field"],
                                $_POST["recaptcha_response_field"]);

if (!$resp->is_valid) {
  die ("The reCAPTCHA wasn't entered correctly. Go back and try it again." .
       "(reCAPTCHA said: " . $resp->error . ")");
}
*/

$naam = $_GET['naam'];
$email = $_GET['email'];
$tekst = $_GET['tekst'];
$val = $_GET['validate'];
$REMOTE_ADDR = $_SERVER['REMOTE_ADDR'];

$banlist = array("81.95.", "91.76.14", "85.140.23", "24.136.65.148", "87.118.106.41", "140.127.139.", "17.149.4.", "80.58.205.", "84.19.176.", "69.147.76", "204.39.95", "203.160.1.", "76.197.155", "200.21.238", "67.175.139", "208.179.21.", "125.46.61.", "210.176.2.", "122.203.23.");

for ($i = 0; $i < sizeof($banlist); $i++)
{
	if (strstr($REMOTE_ADDR, $banlist[$i]))
		die ("You have been banned from posting!");
}

// Check required fields
if ($naam  == "" ||
    $tekst == "") {
	die ("please fill out all fields");
}

if ($email)
if (!is_valid_email($email)) {
    die ("$email is not a valid emailaddress");
}

$naam  = htmlspecialchars(strip_tags(stripslashes($naam)));
$email = htmlspecialchars(strip_tags(stripslashes($email)));
$tekst = htmlspecialchars(strip_tags(stripslashes($tekst)));

add_message($XML_INPUTFILE, $naam, $email, $tekst);

if ($EMAIL_WEBMASTER == "true") { 
	$message = @fdate()." ip: ".$REMOTE_ADDR."\nAuteur: ".$naam."\n".$tekst; 
	mail($EMAILADDRESS_WEBMASTER, "XMLboard addition to wedding site" , $message, "From: ".$email);
}
echo "Message posted!";

/* add_message($infile, $author, $email, $text)
   all args are string.

   adds a new message to the spcified xml file.
*/

function add_message($file, $author, $email, $text) {
	global $REMOTE_ADDR,
           $LOCALE;

	$infile  = fopen("../".$file, "r");
    if(file_exists("../".$file.".out")) {
        die ("filelocking problem, please try again");
    }
	$outfile = fopen("../".$file.".out", "w");
	

	while(!feof($infile)) {
		$line = fgets($infile, 4096); 
		if (!isset($MESSAGE_ADDED) & strtoupper(substr(ltrim($line),0,9)) ==  "<MESSAGE>") {
			fputs($outfile, "<MESSAGE>\n");
			fputs($outfile, "<identifier>(" .@fdate($LOCALE).")</identifier>\n");
			fputs($outfile, "<author email='" .$email. "'>" .$author. "</author>\n");
			fputs($outfile, "<text>" .$text. "</text>\n");
			fputs($outfile, "</MESSAGE>\n\n"); 
			fputs($outfile, $line);

			$MESSAGE_ADDED = true;
		} else {
			fputs($outfile, $line);
		}
	}
	
	fclose($outfile);
	fclose($infile);
	copy("../".$file.".out", "../".$file);
    unlink("../".$file.".out");
}

function fdate($country){
	if ($country == "") {
		$country = "en_US";
	}
	setlocale("LC_ALL", $country); 
	$time = strtolower(strftime("%d %B %Y, %H.%M"));
	return $time;
}

function is_valid_email ($address) {
    return (preg_match(
        '/^[-!#$%&\'*+\\.\/0-9=?A-Z^_`{|}~]+'.   // the user name
        '@'.                                     // the ubiquitous at-sign
        '([-0-9A-Z]+\.)+' .                      // host, sub-, and domain names
        '([0-9A-Z]){2,4}$/i',                    // top-level domain (TLD)
        trim($address)));
}
?>


board/scripts/conf.php
<?php
/****************************************
* XMB v1.0 (XmlMessageBoard)
* http://xmboard.sourceforge.net
*****************************************
* written by erwin vrolijk,
* aierwin@users.sourceforge.net
*****************************************
* this script is released under the GPL
* http://www.gnu.org/copyleft/gpl.html
****************************************/
	$OUTPUT_LINE_WIDTH      = 50;
	$MESSAGES_PER_PAGE      = 20;

	$BASEFILE               = "../../board.html";
 	$XML_INPUTFILE          = "./data/board.xml";
	$XML_OUTPUTFILE         = "./data/board.xml.out";

	$EMAIL_WEBMASTER        = "true"; //if 'true'_the script mails a copy of every writeup to $EMAILADDRESS_WEBMASTER
	$EMAILADDRESS_WEBMASTER = "liu.isaac@gmail.com";
 
    $LOCALE                 = "en_US" //for english/american date format
//  $LOCALE                 = "nl_NL" //for dutch date format
?>


board/scripts/print.php
<?php
/****************************************
* XMB v1.0 (XmlMessageBoard)
* http://xmboard.sourceforge.net
*****************************************
* written by erwin vrolijk,
* aierwin@users.sourceforge.net
*****************************************
* this script is released under the GPL
* http://www.gnu.org/copyleft/gpl.html
****************************************/

/* print.php

   requires conf.php

this file containts print functions
*/

//set global array 
$temp;

/* print_from_file($file)
   string $file will be opened en be print out,
   returns false if $file cannot be opened.
*/
function print_from_file($file) {
	$infile = @fopen ($file, "r");
	if (!$infile) {
		return false;
	}
	
	while(!feof($infile)) {
		$line = fgets($infile, 4096); 
		print($line);
	}
	fclose($infile);
}

/* print_xml_starttag($tag, $attrs)
   string $tag
   array $attrs

   transforms XML to HTML and prints
*/
function print_xml_starttag($tag, $attrs) {
	global $temp;

	if ($tag == "MESSAGE") {
		print("<p id='boardContent'>");
	}
	
	if ($tag == "IDENTIFIER") {
		print("<div id='boardDate'>");
	}

	if ($tag == "AUTHOR") {
		if ($attrs[EMAIL] != "") {
			print("<div id='boardAuthor'>[[");
			// remember we have opened a <A> tag so we can clos it later
			$temp[OPEN_A_TAG] = $attrs[EMAIL];
		}else{
			print("<div id='boardAuthor>");
		}
	}

	if ($tag == "TEXT") { 
		// remember we are in the TEXT-tag
	        print("<div id='boardText'>");
		$temp[IN_TEXT_TAG] = true;
	}
}
	
/* print_xml_endtag($tag, $attrs)
   string $tag
   array $attrs

   transforms XML to HTML and prints
*/	
function print_xml_endtag($tag, $attrs) {
	global $temp;


	if ($tag == "MESSAGE") {
		print("</p>");
	}
	
	if ($tag == "IDENTIFIER") {
		print("</div>");
	}

	if ($tag == "AUTHOR") {
		if (isset($temp[OPEN_A_TAG])) {
			print("| $temp[OPEN_A_TAG] ]]</div>");
			unset($temp[OPEN_A_TAG]);
		}else{
			print("</div>");
		}
	}

	if ($tag == "TEXT") {
		print("</div>");
		unset($temp[IN_TEXT_TAG]);
	}
}


/* print_xml_characterdata($data)
   string $data

   formats and prints $data
*/	
function print_xml_characterdata($data) {
	global $temp;
	global $OUTPUT_LINE_WIDTH;

	if (isset($temp[IN_TEXT_TAG])) {
		print(wordwrap(nl2br($data), $OUTPUT_LINE_WIDTH, '<BR>', 1));
	} else {
		print($data);
	}
}

/* print_redirect($location)

   prints HTML-headers to redirect browser to $location
*/
function print_redirect($location) {
	header ("Location: ".$location);
	exit;
}

/* print_message_navigation($start, $messages_per_page)

   prints line which indicates which messages are displayed,
   and generate links to nex and previous pages
*/
function print_message_navigation($start, $messages_per_page){
	global $BASEFILE;

	print("<P>message ".$start." to ".($start+$messages_per_page)." ");
	if($start > 1) {
		print("<a href='".$BASEFILE."?start=".($start-$messages_per_page)."'>&lt;- newer</a> ");
	}
	print("<a href='".$BASEFILE."?start=".($start+$messages_per_page)."'>older -&gt;</a></P>");
}
?>


board/scripts/xmlparser.php
<?php
/****************************************
* XMB v1.0 (XmlMessageBoard)
* http://xmboard.sourceforge.net
*****************************************
* written by erwin vrolijk,
* aierwin@users.sourceforge.net
*****************************************
* this script is released under the GPL
* http://www.gnu.org/copyleft/gpl.html
****************************************/
/* xmlparser.php

   requires print.php
*/

$message_id    = 0;
$START_MESSAGE = 0;
$END_MESSAGE   = 0;

/* parse_xml($file)
   string $file will be opened and parsed
*/
function parse_xml($file) {
	if (!($fp = @fopen($file, "r"))) {
		die("/scripts/xmlparser.php function:parse_xml could not open XML input");
	}
	
	$xml_parser = xml_parser_create();
	// use case-folding so we are sure to find the tag
	xml_parser_set_option($xml_parser, XML_OPTION_CASE_FOLDING, true);
	xml_set_element_handler($xml_parser, "start_element", "end_element");
	xml_set_character_data_handler($xml_parser, "character_data");

	while ($data = fread($fp, 4096)) {
		if (!xml_parse($xml_parser, $data, feof($fp))) {
			die(sprintf("XML error: %s at line %d",
				     xml_error_string(xml_get_error_code($xml_parser)),
				     xml_get_current_line_number($xml_parser)));
		}
	}
	
	xml_parser_free($xml_parser);
}

function start_element($parser, $name, $attrs) {
	print_xml_starttag($name, $attrs);
}

function end_element($parser, $name) {
	print_xml_endtag($name, $attrs);
}

function character_data($parser, $data) {
	print_xml_characterdata($data);
}

/*read_messages($file, $start_message, $number_of_messages)
  string $file
  int    $start_message
  int    $number_of_messages

  parses xml-file $file for messages, start output at $start_message,
  displays $number_of_messages including $start_message.

  last added message has id=1
*/
function read_messages($file, $start_message, $number_of_messages) {
	global $START_MESSAGE,
	       $END_MESSAGE;

	$START_MESSAGE = $start_message;
	$END_MESSAGE   = $start_message + $number_of_messages;	
	
	if (!($fp = @fopen($file, "r"))) {
		die("/scripts/xmlparser.php function:read_messages could not open XML input");
	}
	
	$xml_parser = xml_parser_create();
	// use case-folding so we are sure to find the tag
	xml_parser_set_option($xml_parser, XML_OPTION_CASE_FOLDING, true);
	xml_set_element_handler($xml_parser, "start_element2", "end_element2");
	xml_set_character_data_handler($xml_parser, "character_data2");

	while ($data = fread($fp, 4096)) {
		if (!xml_parse($xml_parser, $data, feof($fp))) {
			die(sprintf("XML error: %s at line %d",
				     xml_error_string(xml_get_error_code($xml_parser)),
				     xml_get_current_line_number($xml_parser)));
		}
	}
	
	xml_parser_free($xml_parser);
}

function start_element2($parser, $name, $attrs) {
	global $message_id,
	       $START_MESSAGE,
	       $END_MESSAGE;
	
	if ($name == "MESSAGE") {
		$message_id++;
	}
	if ($START_MESSAGE <= $message_id && $message_id <= $END_MESSAGE) {
		print_xml_starttag($name, $attrs);
	}

}

function end_element2($parser, $name) {
	global $message_id,
	       $START_MESSAGE,
	       $END_MESSAGE;
	
	if ($START_MESSAGE <= $message_id && $message_id < $END_MESSAGE) {
		print_xml_endtag($name, $attrs);
	}
}

function character_data2($parser, $data) {
	global $message_id,
	       $START_MESSAGE,
	       $END_MESSAGE;
	
	if ($START_MESSAGE <= $message_id && $message_id < $END_MESSAGE) {
		print_xml_characterdata($data);
	}
}


?>


ajax_mail.js

var xmlHttp;
var text;
var textb;
var areq;

function importXMLDoc() {

    if ( document.getElementById("boardstatus")  ) {
        if ( document.getElementById("boardstatus").hasChildNodes() )
            {
                document.getElementById("boardstatus").removeChild(textb);
            }
        textb=document.createTextNode("Refreshing board....");
        document.getElementById("boardstatus").appendChild(textb);
    }
    areq=GetXmlHttpObject();
    if (areq==null)
        {
        if (document.getElementById("boardstatus")  ) {
            if (  document.getElementById("boardstatus").hasChildNodes() )
                {
                    document.getElementById("boardstatus").removeChild(textb);
                }
            textb=document.createTextNode("Your browser doesn't support this script. Please use IE/firefox/Safari.");
            document.getElementById("boardstatus").appendChild(textb);
	} 
      }
      else
      {
        areq.onreadystatechange = processReqChange;
        areq.open("GET", "board/board.php", true);
        areq.send(null);


      }
}


// handle onreadystatechange event of req object
function processReqChange() {
    // only if req shows "loaded"
    if (areq.readyState == 4) {
        // only if "OK"
        if (areq.status == 200) {
            createMain();
            if (document.getElementById("boardstatus") ) {
                if (  document.getElementById("boardstatus").hasChildNodes() )
                    {
                        document.getElementById("boardstatus").removeChild(textb);
                    }
		var currentTime = new Date();
		var month = currentTime.getMonth() + 1;
		var year = currentTime.getFullYear();
		var day = currentTime.getDate(); 
		var min = currentTime.getMinutes();
		var sec = currentTime.getSeconds();
		var hour = currentTime.getHours();
		var ampm;
		if (min < 10)
			min = "0"+min;
		if (sec < 10 )
			sec = "0"+sec;
		if ( hour <= 12 ){
			ampm = "am";
                        if (hour == 0)
				hour = 12;
		}
		else{
			ampm = "pm";
			hour = hour - 12;
		}	
                textb=document.createTextNode(month+"/"+day+" "+hour+":"+min+":"+sec+" "+ampm);
                document.getElementById("boardstatus").appendChild(textb);
            }
        } else {
            alert("There was a problem retrieving the XML data:\n" +
                  areq.statusText);
        }
    }
}
 
function createMain()
{
	var xb = areq.responseXML.getElementsByTagName('MESSAGE');
	var para = document.createElement('div');
    para.setAttribute('id', 'boardContent');
	var len = xb.length<20 ? xb.length:20;
    for (i=0; i<len; i++)
        {
	
            var j = 0;
		
            while (xb[i].childNodes[j].nodeType !=1) j++;
            var theDate = document.createTextNode(xb[i].childNodes[j].firstChild.nodeValue);
            j++;
            while (xb[i].childNodes[j].nodeType !=1) j++;
            var author = document.createTextNode(xb[i].childNodes[j].firstChild.nodeValue);
            var authemail = "mailto:";
            authemail += xb[i].childNodes[j].attributes.getNamedItem('email').nodeValue;
            j++;
            while (xb[i].childNodes[j].nodeType !=1) j++;
            var msgtext = document.createTextNode(xb[i].childNodes[j].firstChild.nodeValue);
		
		
            var newEl = document.createElement('TABLE');
            newEl.setAttribute('cellPadding',1);
            newEl.setAttribute('width', '95%');
            var tmp = document.createElement('TBODY');
            newEl.appendChild(tmp);
            para.appendChild(newEl);
            var row = document.createElement('TR');
            var container = document.createElement('TH');
            container.appendChild(author);
            row.appendChild(container);
            var container2 = document.createElement('TD');
            container2.appendChild(theDate);
            container2.setAttribute('id', 'msgBoardDate');
            container2.setAttribute('align', 'right');
            row.appendChild(container2);
            tmp.appendChild(row);

            var row2 = document.createElement('TR');
            var container3 = document.createElement('TD');
            container3.setAttribute('ColSpan', '2');
            container3.appendChild(msgtext);
            row2.appendChild(container3);
            tmp.appendChild(row2);
            para.appendChild(newEl);

        } 

    if ( document.getElementById('maincontent')) {
        if ( document.getElementById('maincontent').hasChildNodes() ) {
            document.getElementById('maincontent').replaceChild(para, document.getElementById('maincontent').childNodes[0]);	
        }
        else
            document.getElementById('maincontent').appendChild(para);	
    }
}



    
function mailform()
{
    xmlHttp=GetXmlHttpObject();
    if (xmlHttp==null)
        {
            alert ("Browser does not support HTTP Request");
            return
                } 
    if ( document.getElementById("formstatus") ) {
        if (  document.getElementById("formstatus").hasChildNodes() )
            {
                document.getElementById("formstatus").removeChild(text);
            }
        text=document.createTextNode("Sending....");
        document.getElementById("formstatus").appendChild(text);
    }
                    
    var name = document.email_form.yourname.value;
    var email= document.email_form.email_address.value;
    var comments = document.email_form.comments.value;
    var url ="board/scripts/add.php";
    url= url+"?naam="+name+"&email="+email+"&tekst="+comments;	
    xmlHttp.onreadystatechange=stateChanged;
    xmlHttp.open("GET",url,true);
    xmlHttp.send(null);
} 
                                                        
                                                            
function stateChanged() 
{ 
    if (xmlHttp.readyState==4 || xmlHttp.readyState=="complete")
        { 
            if (xmlHttp.status==200)
                {
                    document.email_form.yourname.value="";
                    document.email_form.email_address.value="";
                    document.email_form.comments.value="";
                    if ( document.getElementById("formstatus") ) {
                        if ( document.getElementById("formstatus").hasChildNodes() )
                            {
                                document.getElementById("formstatus").removeChild(text);
                            }
                        text=document.createTextNode(xmlHttp.responseText);
                        document.getElementById("formstatus").appendChild(text);
                    }
                }
        } 
    importXMLDoc();
} 
    
function GetXmlHttpObject() { 
    var objXMLHttp=null;
    if (window.XMLHttpRequest)
        {
            objXMLHttp=new XMLHttpRequest();
        }
    else if (window.ActiveXObject)
        {
            objXMLHttp=new ActiveXObject("Microsoft.XMLHTTP");

        }
    return objXMLHttp;
} 

importXMLDoc();
